pkg/object/meshcontroller/spec/spec.go

```diff
@@ -629,114 +672,144 @@ func (b *pipelineSpecBuilder) appendTimeLimiter(tl *timelimiter.Spec) *pipelineS
 	return b
 }
 
-func (b *pipelineSpecBuilder) appendProxyWithCanary(instanceSpecs []*ServiceInstanceSpec, canary *Canary, lb *proxy.LoadBalance, cert, rootCert *Certificate) *pipelineSpecBuilder {
-	mainServers := []*proxy.Server{}
-	canaryInstances := []*ServiceInstanceSpec{}
+func (b *pipelineSpecBuilder) appendProxyWithCanary(instanceSpecs []*ServiceInstanceSpec,
+	canaries []*ServiceCanary, lb *proxy.LoadBalance, cert, rootCert *Certificate) *pipelineSpecBuilder {
+
+	filterName := "backend"
+	if lb == nil {
+		lb = &proxy.LoadBalance{
+			Policy: proxy.PolicyRoundRobin,
+		}
+	}
 
 	needMTLS := false
 	if cert != nil && rootCert != nil {
 		needMTLS = true
 	}
 
-	for k, instanceSpec := range instanceSpecs {
-		if instanceSpec.Status == ServiceStatusUp {
-			if len(instanceSpec.Labels) == 0 {
-				if needMTLS {
-					mainServers = append(mainServers, &proxy.Server{
-						URL: fmt.Sprintf("https://%s:%d", instanceSpec.IP, instanceSpec.Port),
-					})
-				} else {
-					mainServers = append(mainServers, &proxy.Server{
-						URL: fmt.Sprintf("http://%s:%d", instanceSpec.IP, instanceSpec.Port),
-					})
+	mainPool := &proxy.PoolSpec{
+		LoadBalance: lb,
+	}
+	candidatePools := make([]*proxy.PoolSpec, len(canaries))
 
-				}
-			} else {
-				canaryInstances = append(canaryInstances, instanceSpecs[k])
-			}
+	filter := map[string]interface{}{
+		"kind":     proxy.Kind,
+		"name":     filterName,
+		"mainPool": mainPool,
+	}
+	if needMTLS {
+		filter["mtls"] = &proxy.MTLS{
+			CertBase64:     cert.CertBase64,
+			KeyBase64:      cert.KeyBase64,
+			RootCertBase64: rootCert.CertBase64,
 		}
 	}
-	backendName := "backend"
 
-	if lb == nil {
-		lb = &proxy.LoadBalance{
-			Policy: proxy.PolicyRoundRobin,
+	makeServer := func(instance *ServiceInstanceSpec) *proxy.Server {
+		var url string
+		if needMTLS {
+			url = fmt.Sprintf("https://%s:%d", instance.IP, instance.Port)
+		} else {
+			url = fmt.Sprintf("http://%s:%d", instance.IP, instance.Port)
```

Comment:
Nit: only difference between line 711 and 713 is `s`, maybe this if-else could select protocol and then construct url like `fmt.Sprintf("%s://%s:%d", protocol, instance.IP, instance.Port)`